File: src/main/java/com/rs/game/content/minigames/castlewars/CastleWarsWaitingController.java
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//  Copyright (C) 2021 Trenton Kress
//  This file is part of project: Darkan
//
package com.rs.game.content.minigames.castlewars;

import com.rs.game.model.entity.player.Controller;
import com.rs.game.model.entity.player.Equipment;
import com.rs.game.model.object.GameObject;
import com.rs.lib.game.WorldTile;
import com.rs.lib.net.ClientPacket;

public class CastleWarsWaitingController extends Controller {

	private int team;

	public CastleWarsWaitingController(int team) {
		this.team = team;
	}

	@Override
	public void start() {
		sendInterfaces();
	}

	// You can't leave just like that!

	public void leave() {
		player.getInterfaceManager().removeOverlay();
		CastleWars.removeWaitingPlayer(player, team);
	}

	@Override
	public void sendInterfaces() {
		player.getInterfaceManager().sendOverlay(57);
	}

	@Override
	public boolean processButtonClick(int interfaceId, int componentId, int slotId, int slotId2, ClientPacket packet) {
		if (interfaceId == 387)
			if (componentId == 9 || componentId == 6) {
				player.sendMessage("You can't remove your team's colours.");
				return false;
			}
		return true;
	}

	@Override
	public boolean canEquip(int slotId, int itemId) {
		if (slotId == Equipment.CAPE || slotId == Equipment.HEAD) {
			player.sendMessage("You can't remove your team's colours.");
			return false;
		}
		return true;
	}

	@Override
	public boolean sendDeath() {
		removeController();
		leave();
		return true;
	}

	@Override
	public boolean logout() {
		player.setTile(WorldTile.of(CastleWars.LOBBY, 2));
		return true;
	}

	@Override
	public boolean processMagicTeleport(WorldTile toTile) {
		player.simpleDialogue("You can't leave just like that!");
		return false;
	}

	@Override
	public boolean processItemTeleport(WorldTile toTile) {
		player.simpleDialogue("You can't leave just like that!");
		return false;
	}

	@Override
	public boolean processObjectTeleport(WorldTile toTile) {
		player.simpleDialogue("You can't leave just like that!");
		return false;
	}

	@Override
	public boolean processObjectClick1(GameObject object) {
		int id = object.getId();
		if (id == 4389 || id == 4390) {
			removeController();
			leave();
			return false;
		}
		return true;
	}

	@Override
	public void magicTeleported(int type) {
		removeController();
		leave();
	}

	@Override
	public void forceClose() {
		leave();
	}
}

File: src/main/java/com/rs/game/content/minigames/castlewars/CastleWarsWaitingController.kt
package com.rs.game.content.minigames.castlewars

import com.rs.game.content.skills.magic.TeleType
import com.rs.game.model.entity.Teleport
import com.rs.game.model.entity.player.Controller
import com.rs.game.model.entity.player.Equipment
import com.rs.game.model.`object`.GameObject
import com.rs.lib.game.Tile
import com.rs.lib.net.ClientPacket

class CastleWarsWaitingController(private val team: Int) : Controller() {

    override fun start() {
        sendInterfaces()
    }

    fun leave() {
        player.interfaceManager.removeOverlay()
        CastleWars.removeWaitingPlayer(player, team)
    }

    override fun sendInterfaces() {
        player.interfaceManager.sendOverlay(57)
    }

    override fun processButtonClick(interfaceId: Int, componentId: Int, slotId: Int, slotId2: Int, packet: ClientPacket?): Boolean {
        when (interfaceId) {
            387 -> {
                if (componentId == 9 || componentId == 6)
                    player.sendMessage("You can't remove your team's colours.")
                return false
            }
        }
        return true
    }

    override fun canEquip(slotId: Int, itemId: Int): Boolean {
        if (slotId == Equipment.CAPE || slotId == Equipment.HEAD) {
            player.sendMessage("You can't remove your team's colours.")
            return false
        }
        return true
    }

    override fun sendDeath(): Boolean {
        removeController()
        leave()
        return true
    }

    override fun logout(): Boolean {
        player.tile = Tile.of(CastleWars.LOBBY, 2)
        return true
    }

    override fun processTeleport(tele: Teleport?): Boolean {
        player.simpleDialogue("You can't leave just like that!")
        return false
    }

    override fun processObjectClick1(obj: GameObject?): Boolean {
        val id = obj?.id
        if (id == 4389 || id == 4390) {
            removeController()
            leave()
            return false
        }
        return true
    }

    override fun onTeleported(type: TeleType?) {
        removeController()
        leave()
    }

    override fun forceClose() {
        leave()
    }

}